Test/Microsoft.ApplicationInsights.Test/Shared/Extensibility/Implementation/Tracing/HeartbeatTests.cs

```diff
@@ -0,0 +1,357 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Linq;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.Mocks;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Microsoft.ApplicationInsights.TestFramework;
+
+    [TestClass]
```

Comment:
I'd strongly suggest a test covering this end to end.
//Arrange
Initialize DiagnosticTelemetryModule with a channel which keeps items sent in memory, and heartbeart interval set to 100msec.

//Act
Sleep >300 msec, so that atleast 3 hearbeats will be sent.

//Validate
Validate that 3 items are sent.

Reply:
I disagree, as using Timer callbacks in unit tests can lead to false negatives, which leads to unnecessary investigation time. Perhaps instead of testing System.Threading.Timer we should keep the units of functionality tested in as modular manner possible? I fully agree that this code should be functionally tested when we get it to dotnet-server though, where we can add a functional e2e test that can wait for a time period instead. Is there any plan to enable functional testing for this repo?

Reply:
1. With hbinterval set to very low like 10ms, and sleeping for, say, 300msec,  unit test can check if atleast 1 or 2 hb was sent, not exactly 300/10=30 items. (Correct me if  this is also risky/unstable - i believe this kind of logic already exists in this repo for sampling/server channel etc).
2. its not a good idea to move the testing to dotnet-server repo. (also, some info about functional tests in server repo - the fire-request->wait()->lookforevent mechanism used in server functional tests are NOT stable as we see random failures all the time. SDK may not flush the items in time is cause for this. And that' why i suggested to use use mock inmemory channel to test this.)

If unit tests are fully covering the functionality, then its okay to not have this e2e test, but I'll let Dmitry/Sergey also to share their thoughts on testing e2e.

Reply:
1. It does seem like it could work, but it's possible that it will fail (even for large sample times, and small expectations it still *can* and therefore *will* fail). However, if we feel strongly enough about this I can put the test in, if the developer cost to investigate these failures is minimal (I would put a comment in the test indicating that this *can* fail due to timing mishaps/test machine being overloaded to hopefully offset this potential downside).

2. The unreliability of integration/functional tests is pretty common. My comment was based on my experience here: keep the unreliable tests together, and cope with them in a consistent manner. Whereas keeping unit tests as stable as possible is typically my goal with the tests we see here. Again, I am happy to give up a bit of unreliability if the team is ok with it.